
**Comentario del Revisor**

Hola!

Soy Juan Manuel Romero, pero siéntete libre de llamarme Juanma. Soy code reviewer en Tripleten y hoy estaré revisando tu entrega.

Para simular la dinámica de un ambiente de trabajo, si veo algún error, en primer instancia solo los señalaré, dándote la oportunidad de encontrarlos y corregirlos por tu cuenta. En un trabajo real, el líder de tu equipo hará una dinámica similar. En caso de que no puedas resolver la tarea, te daré una información más precisa en la próxima revisión. 

Solo un aviso rápido: cuando estés revisando el proyecto, por favor deja mis comentarios originales tal como están. De esta manera, podemos seguir fácilmente el progreso y asegurarnos de que no se nos pase nada por alto. Y, si realizas algún cambio basado en mis comentarios, sería genial si pudieras resaltar esas actualizaciones para que se destaquen.

Puedes encontrar mis comentarios en cajas verdes, amarillas o rojas como estas:

<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor</b> <a class="tocSkip"></a>

Éxito. Todo se ha hecho correctamente.

</div>


<div class="alert alert-block alert-warning"> 
<b>Comentario del Revisor</b> <a class="tocSkip"></a>

Observaciones. Algunas recomendaciones.

</div> 


<div class="alert alert-block alert-danger">
<b>Comentario del Revisor</b> <a class="tocSkip"></a>

Requiere corrección. El bloque requiere algunas correcciones. El trabajo no puede ser aceptado con los comentarios en rojo.

</div>

Puedes responderme usando esto:

<div class="alert alert-block alert-info"> <b>Respuesta del estudiante.</b> <a class="tocSkip"></a> </div>

Proyecto Machine Learning

1. Carga y examinación del archivo de datos

In [1]:
import pandas as pd

# Ruta del archivo
#file_path = '/home/jovyan/work/users_behavior.csv'
file_path = "/datasets/users_behavior.csv"

# Cargar el archivo
data = pd.read_csv(file_path)

Carga de librerías

In [2]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report
import matplotlib.pyplot as plt
import seaborn as sns

Exploración del dataset

In [3]:
print(data.head())
print(data.info())
print(data.describe())

   calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB
None
             calls      minutes     messages       mb_used     is_ultra
count  3214.000000  3214.000000  3214.000000   3214.000000  3214.000000
mean     63.038892   438.208787    38.281269  17207.673836     0.306472
std      33.236368   234.569872    36.148326   7570.968246  

<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Buen trabajo! Has utilizado métodos como _info()_, _head()_ o _describe()_ para analizar los datos. Estos métodos te permiten entender la estructura de los datos, los tipos de variables, la presencia de valores faltantes y obtener una descripción estadística general.

</div>

2. Segmentación de los datos fuente en un conjunto de entrenamiento, uno de validación y uno de prueba.

Conjunto de entrenamiento: 60%
Conjunto de validación: 20%
Conjunto de prueba: 20%

In [4]:
from sklearn.model_selection import train_test_split

# Separar las características (X) y el objetivo (y)
X = data.drop(columns=['is_ultra'])
y = data['is_ultra']

# Dividir el dataset en entrenamiento + validación (80%) y prueba (20%)
X_train_val, X_test, y_train_val, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# Dividir el conjunto de entrenamiento + validación en entrenamiento (60%) y validación (20%)
X_train, X_val, y_train, y_val = train_test_split(
    X_train_val, y_train_val, test_size=0.25, random_state=42  # 0.25 x 0.8 = 0.2 (20%)
)

# Confirmar tamaños de los conjuntos
print("Tamaños de los conjuntos:")
print(f"Entrenamiento: {X_train.shape[0]} muestras")
print(f"Validación: {X_val.shape[0]} muestras")
print(f"Prueba: {X_test.shape[0]} muestras")

Tamaños de los conjuntos:
Entrenamiento: 1928 muestras
Validación: 643 muestras
Prueba: 643 muestras


<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Excelente! Has dividido correctamente los datos en tres conjuntos: _train_, _validation_ y _test_. Dividir los datos de esta forma permite entrenar el modelo en un conjunto, validar su desempeño en otro y finalmente evaluarlo en un conjunto independiente para obtener una medida precisa de su rendimiento. 
    
Además, has utilizado correctamente _train_test_split_ dos veces para dividir los datos en tres conjuntos (entrenamiento, validación y prueba). Este método es el más común para dividir los datos de manera aleatoria, asegurando que cada conjunto tenga una representación justa.

</div>

3. Investigación de la calidad de diferentes modelos cambiando los hiperparámetros. Descripción breve de los hallazgos del estudio.

<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Muy bien hecho! Has ajustado correctamente los hiperparámetros, lo que es una excelente práctica. Ajustar los hiperparámetros permite optimizar el modelo para obtener el mejor rendimiento posible.

</div>

Regresión logística:

In [5]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV

# Inicialización del modelo de regresión logística
logreg = LogisticRegression(random_state=42)

# Hiperparámetros para ajustar con GridSearchCV
param_grid = {
    'penalty': ['l1', 'l2'],  # Regularización L1 y L2
    'C': [0.01, 0.1, 1, 10],  # Inverso de la regularización
    'solver': ['liblinear'],  # Solver adecuado para L1
}

# Ajuste de hiperparámetros con GridSearchCV
grid_search = GridSearchCV(estimator=logreg, param_grid=param_grid, cv=5, n_jobs=-1, verbose=1)
grid_search.fit(X_train, y_train)

# Mejor conjunto de hiperparámetros
print("Mejores hiperparámetros encontrados:", grid_search.best_params_)

# Evaluación el modelo en el conjunto de validación
y_pred_val = grid_search.predict(X_val)
accuracy_val = accuracy_score(y_val, y_pred_val)
print(f"Exactitud en el conjunto de validación: {accuracy_val:.4f}")

Fitting 5 folds for each of 8 candidates, totalling 40 fits
Mejores hiperparámetros encontrados: {'C': 1, 'penalty': 'l1', 'solver': 'liblinear'}
Exactitud en el conjunto de validación: 0.7434


Árbol de decisión:

In [6]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV  # Importación necesaria
from sklearn.metrics import accuracy_score

# Inicialización del modelo de Árbol de Decisión
dt = DecisionTreeClassifier(random_state=42)

# Hiperparámetros para ajustar con GridSearchCV
param_grid_dt = {
    'criterion': ['gini', 'entropy'],  # Función de evaluación
    'max_depth': [None, 10, 20, 30],  # Profundidad máxima
    'min_samples_split': [2, 5, 10],  # Mínimo de muestras para dividir un nodo
}

# Ajuste de hiperparámetros con GridSearchCV
grid_search_dt = GridSearchCV(estimator=dt, param_grid=param_grid_dt, cv=5, n_jobs=-1, verbose=1)
grid_search_dt.fit(X_train, y_train)

# Mejor conjunto de hiperparámetros
print("Mejores hiperparámetros Árbol de Decisión:", grid_search_dt.best_params_)

# Evaluación en el conjunto de validación
y_pred_val_dt = grid_search_dt.predict(X_val)
accuracy_val_dt = accuracy_score(y_val, y_pred_val_dt)
print(f"Exactitud en el conjunto de validación (Árbol de Decisión): {accuracy_val_dt:.4f}")

Fitting 5 folds for each of 24 candidates, totalling 120 fits
Mejores hiperparámetros Árbol de Decisión: {'criterion': 'entropy', 'max_depth': 10, 'min_samples_split': 5}
Exactitud en el conjunto de validación (Árbol de Decisión): 0.7574


Random forest:

In [9]:
from sklearn.ensemble import RandomForestClassifier

# Inicialización del modelo de Random Forest
rf = RandomForestClassifier(random_state=42)

# Hiperparámetros para ajustar con GridSearchCV
param_grid_rf = {
    'n_estimators': [50, 100, 200],  # Número de árboles
    'max_depth': [None, 10, 20],  # Profundidad máxima
    'min_samples_split': [2, 5, 10],  # Mínimo de muestras para dividir un nodo
}

# Ajuste de hiperparámetros con GridSearchCV
grid_search_rf = GridSearchCV(estimator=rf, param_grid=param_grid_rf, cv=5, n_jobs=-1, verbose=1)
grid_search_rf.fit(X_train, y_train)

# Mejor conjunto de hiperparámetros
print("Mejores hiperparámetros Random Forest:", grid_search_rf.best_params_)

# Evaluación en el conjunto de validación
y_pred_val_rf = grid_search_rf.predict(X_val)
accuracy_val_rf = accuracy_score(y_val, y_pred_val_rf)
print(f"Exactitud en el conjunto de validación (Random Forest): {accuracy_val_rf:.4f}")

Fitting 5 folds for each of 27 candidates, totalling 135 fits
Mejores hiperparámetros Random Forest: {'max_depth': 10, 'min_samples_split': 2, 'n_estimators': 200}
Exactitud en el conjunto de validación (Random Forest): 0.7932


Comparación de resultados:

In [11]:
results = pd.DataFrame({
    'Modelo': ['Regresión Logística', 'Árbol de Decisión', 'Random Forest'],
    'Exactitud Validación': [accuracy_val, accuracy_val_dt, accuracy_val_rf]
})

print(results)

                Modelo  Exactitud Validación
0  Regresión Logística              0.743390
1    Árbol de Decisión              0.757387
2        Random Forest              0.793157


<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Muy bien! Has entrenado el modelo únicamente con el conjunto de entrenamiento, lo que es exactamente lo que se debe hacer. Este conjunto contiene los datos con los que el modelo aprende los patrones, y no debe utilizar otros conjuntos durante el entrenamiento.

Luego, has validado correctamente el modelo utilizando solo el conjunto de validación. Esta es una práctica esencial para ajustar los hiperparámetros y comparar diferentes modelos sin introducir sesgos

</div>

Análisis de los Resultados:

Regresión Logística (0.7434):
Es un modelo base sencillo que proporciona una referencia inicial.
Aunque es menos preciso que los otros modelos, es útil para interpretar la relación entre variables y la variable objetivo.

Árbol de Decisión (0.7574):
Mejoró ligeramente la exactitud en comparación con la regresión logística.
Este modelo es fácil de interpretar y explicar, pero puede ser propenso al sobreajuste en datos complejos.

Random Forest (0.7932):
Fue el modelo más preciso en validación.
Combina múltiples árboles de decisión para reducir el riesgo de sobreajuste y mejorar la generalización.
Su rendimiento sugiere que las características tienen relaciones complejas que el bosque aleatorio puede capturar.

Hiperparámetros utilizados:

Regresión Logística:
penalty: Tipo de regularización (l1 o l2).
C: Inverso de la fuerza de regularización.
solver: Método de optimización (liblinear).

Árbol de Decisión:
max_depth: Profundidad máxima del árbol.
min_samples_split: Mínimo de muestras necesarias para dividir un nodo.
min_samples_leaf: Mínimo de muestras necesarias en una hoja.

Random Forest:
n_estimators: Número de árboles en el bosque.
max_depth: Profundidad máxima de cada árbol.
min_samples_split: Mínimo de muestras necesarias para dividir un nodo.

<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #2</b> <a class="tocSkip"></a>

Excelentes cambios Emilio! Ahora si ha quedado todo super bien documentado

</div>

4. Comprobación de la calidad del modelo usando el conjunto de prueba.

In [7]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Entrenamiento del modelo con los mejores hiperparámetros encontrados
best_logreg = grid_search.best_estimator_

# Realización de predicciones en el conjunto de prueba
y_pred_test = best_logreg.predict(X_test)

# Cálculo de la exactitud en el conjunto de prueba
accuracy_test = accuracy_score(y_test, y_pred_test)
print(f"Exactitud en el conjunto de prueba: {accuracy_test:.4f}")

# Matriz de confusión
conf_matrix = confusion_matrix(y_test, y_pred_test)
print("Matriz de confusión:")
print(conf_matrix)

# Reporte de clasificación con precisión, recall y F1-score
class_report = classification_report(y_test, y_pred_test)
print("Reporte de clasificación:")
print(class_report)

Exactitud en el conjunto de prueba: 0.7605
Matriz de confusión:
[[439  16]
 [138  50]]
Reporte de clasificación:
              precision    recall  f1-score   support

           0       0.76      0.96      0.85       455
           1       0.76      0.27      0.39       188

    accuracy                           0.76       643
   macro avg       0.76      0.62      0.62       643
weighted avg       0.76      0.76      0.72       643



Exactitud en el conjunto de prueba:

Exactitud: 0.7605: El modelo ha alcanzado una exactitud de 76.05% en el conjunto de prueba, lo que es superior al umbral de 0.75 que se estableció al inicio del proyecto. Esto sugiere que el modelo tiene un buen desempeño general.

Matriz de confusión:
lua

[[439  16]
 [138  50]]
 
439 Verdaderos Negativos (TN): Casos en los que el modelo predijo correctamente la clase "0" (Smart).

50 Verdaderos Positivos (VP): Casos en los que el modelo predijo correctamente la clase "1" (Ultra).

16 Falsos Positivos (FP): Casos en los que el modelo predijo erróneamente la clase "1" cuando en realidad la clase correcta era "0".

138 Falsos Negativos (FN): Casos en los que el modelo predijo erróneamente la clase "0" cuando en realidad la clase correcta era "1".

Reporte de clasificación:

Clase 0 (Smart):

Precisión: 0.76: El modelo tiene un 76% de precisión al predecir la clase 0.

Recall: 0.96: El modelo logra identificar correctamente el 96% de los verdaderos casos de la clase 0.

F1-Score: 0.85: El modelo tiene un buen balance entre precisión y recall para la clase 0.

Clase 1 (Ultra):

Precisión: 0.76: El modelo tiene un 76% de precisión al predecir la clase 1.

Recall: 0.27: El modelo sólo identifica correctamente el 27% de los verdaderos casos de la clase 1.

F1-Score: 0.39: El modelo presenta un bajo desempeño en la clase 1, lo que indica que necesita mejorar en la identificación de los usuarios que deberían estar en el plan Ultra.

Promedio macro y ponderado:

Macro promedio: La precisión es del 76%, pero el recall y F1-score son menores (62% en promedio). Esto indica que el modelo tiene un buen desempeño en términos generales, pero está sesgado hacia la clase 0 (Smart).

Promedio ponderado: Al tener más muestras de la clase 0, las métricas ponderadas reflejan un mejor desempeño general, pero el modelo tiene dificultades para identificar correctamente la clase 1 (Ultra).

Conclusión:

El modelo ha superado el umbral de exactitud de 0.75, con una exactitud global de 0.76. Sin embargo, el rendimiento para la clase 1 (Ultra) es bajo, especialmente en términos de recall (solo 27%). El modelo está sesgado hacia la clase 0, lo que sugiere que podría mejorarse utilizando técnicas como balanceo de clases o ajuste de pesos en el modelo para mejorar la identificación de los usuarios que deben cambiar al plan Ultra.

<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Has documentado los resultados de manera clara, lo que es crucial para comunicar los hallazgos y justificar la elección del mejor modelo. Explicar el razonamiento detrás de tus decisiones es clave para asegurar que los resultados sean interpretados correctamente.

</div>

5. Prueba de cordura del modelo (sanity check) 

Comprobación del modelo base

In [8]:
from sklearn.dummy import DummyClassifier
from sklearn.metrics import accuracy_score

# Modelo base: clasificador que predice siempre la clase mayoritaria
dummy_clf = DummyClassifier(strategy='most_frequent')
dummy_clf.fit(X_train, y_train)

# Realización de predicciones en el conjunto de prueba
y_pred_dummy = dummy_clf.predict(X_test)

# Exactitud del modelo base
accuracy_dummy = accuracy_score(y_test, y_pred_dummy)
print(f"Exactitud del modelo base: {accuracy_dummy:.4f}")

Exactitud del modelo base: 0.7076


La exactitud del modelo base es 0.7076, lo que significa que el modelo que predice siempre la clase mayoritaria (en este caso, "Smart" o clase 0) tiene una exactitud del 70.76%. Este valor indica que el modelo base tiene un desempeño relativamente decente debido a que la clase mayoritaria ("Smart") representa una porción significativa del conjunto de datos.


<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #2</b> <a class="tocSkip"></a>

Felicidades Emilio! Tu trabajo es excelente 

Has completado todos los items necesarios para aprobar la entrega.

</div>